### ISSUE TITLE
torch.stack for sequences

### ISSUE THREAD

**Issue by yueyinqiu** [pytorch/pytorch#144671]
### 🚀 The feature, motivation and pitch

I'm trying to let my tensors to have a static dimension count for type checking, like:

```python
import torch
import typing

Tensor2d = typing.NewType("Tensor2d", torch.Tensor)
def matmul(x: Tensor2d, y: Tensor2d) -> Tensor2d:
    return Tensor2d(x.matmul(y))
# So that we won't pass any other tensors with wrong dims in accident.
```

However, I found that it is impossible to use things like `torch.stack` on a `list[Tensor2d]`:

```python
import torch
import typing

Tensor2d = typing.NewType("Tensor2d", torch.Tensor)
my_list: list[Tensor2d] = []
torch.stack(my_list)
# Pylance for example, says:
# Argument of type "list[Tensor2d]" cannot be assigned to parameter "tensors" of type "Tuple[Tensor, ...] | List[Tensor]" in function "stack"
#   Type "list[Tensor2d]" is not assignable to type "Tuple[Tensor, ...] | List[Tensor]"
#     "list[Tensor2d]" is not assignable to "Tuple[Tensor, ...]"
#     "list[Tensor2d]" is not assignable to "List[Tensor]"
#       Type parameter "_T@list" is invariant, but "Tensor2d" is not the same as "Tensor"
#       Consider switching from "list" to "Sequence" which is covariantPylancereportArgumentType
```

So I wonder if it is possible to switch the signature from

```python
def stack(tensors: Union[Tuple[Tensor, ...], List[Tensor]], dim: _int = 0, *, out: Optional[Tensor] = None) -> Tensor:
```

to

```python
def stack(tensors: Sequence[Tensor], dim: _int = 0, *, out: Optional[Tensor] = None) -> Tensor:
```

I guess there are some optimization measures that can only be applied on `tuple` and `list`, but is it possible to check the type at runtime? And if it's not a `list`, we could automatically convert it.

And same for some other functions like `cat`. Thanks in advance.

**Comment by ezyang** [comment]
Hmm.

Advertising these functions take Sequence is a lie (I don't believe we support arbitrary sequence-like types) but perhaps it is a useful lie (and suggests that maybe we should adjust our bindings so that we do.) The only alternative in Mypy's type system, I think, is to create a type variable for Tensor which is marked covariant and then use that in lieu of Tensor, which sounds annoying.

So maybe I am OK with Sequence. Someone's got to implement it though!

**Comment by yueyinqiu** [comment]
Hmm... I'm glad if I could be of help. But actually I even failed to find the real implemention...

I have found a `stack` method defined in [torch/_refs](https://github.com/pytorch/pytorch/blob/main/torch/_refs/__init__.py#L3936). And it's calling `torch.cat`.

By clicking `cat` here, vscode directed me to a `_VariableFunctions.pyi` file in my site-packages. I don't think it's what I want. So I just searched for `cat` and also found one in [torch/_refs](https://github.com/pytorch/pytorch/blob/main/torch/_refs/__init__.py#L2763).

And this `cat` is calling `prims.cat`, which is in [torch/_prims](https://github.com/pytorch/pytorch/blob/main/torch/_prims/__init__.py#L266), and it's `impl_aten` is `torch.cat(tensors, dim)` [here](https://github.com/pytorch/pytorch/blob/main/torch/_prims/__init__.py#L1801)...

I think I'm really confused...

**Comment by ezyang** [comment]
For the type stubs on torch.cat and friends it's this code

```
    elif isinstance(t, ListType):
        if str(t.elem) == "int":
            ret = "Union[_int, _size]" if t.size is not None else "_size"
        elif t.is_tensor_like():
            # TODO: this doesn't seem right...
            # Tensor?[] currently translates to Optional[Union[tuple[Tensor, ...], list[Tensor]]]
            # It should probably translate to   Union[tuple[Optional[Tensor], ...], list[Optional[Tensor]]]
            add_optional = True
            ret = (
                "Union[Tensor, tuple[Tensor, ...], list[Tensor]]"
                if t.size is not None
                else "Union[tuple[Tensor, ...], list[Tensor]]"
```

in `torchgen/api/python.py`. You'll then need to do a build to regenerate the type stubs. Maybe `.ci/pytorch/codegen-test.sh` still works.

**Comment by yueyinqiu** [comment]
Hmm... And the real implemention should also be checked right? Since for `float`s we do use `Sequence[_float]`, there must be some problems if we just change the type stubs I think... Maybe there should be someone more familiar with this to check it later.

By the way, a current workaround in my case could be `typing.cast`.

**Comment by ezyang** [comment]
These type stubs don't affect runtime behavior at all

**Comment by yueyinqiu** [comment]
yes, but thats the point. that's really cheating if we just change the type stubs without checking whether the real implemention supports all sequences.
